src/Tools/dotnet-monitor/DiagnosticLifetimeBackgroundService.cs

```diff
@@ -0,0 +1,64 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using Microsoft.Diagnostics.Monitoring.WebApi;
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.Diagnostics.Tools.Monitor
+{
+    /// <summary>
+    /// Base class for implementing a long running <see cref="IDiagnosticLifetimeService"/>.
+    /// </summary>
+    /// <remarks>
+    /// Similar to <see cref="Microsoft.Extensions.Hosting.BackgroundService"/>.
+    /// </remarks>
+    internal abstract class DiagnosticLifetimeBackgroundService :
+        IDiagnosticLifetimeService,
+        IAsyncDisposable
+    {
+        private Task _executingTask;
+        private CancellationTokenSource _stoppingSource;
+
+        public virtual ValueTask DisposeAsync()
+        {
+            _stoppingSource?.SafeCancel();
+
+            return ValueTask.CompletedTask;
+        }
+
+        public async ValueTask StartAsync(CancellationToken cancellationToken)
+        {
+            _stoppingSource = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken);
+
+            // Begin to execute but don't wait for it to complete.
+            _executingTask = ExecuteAsync(_stoppingSource.Token);
```

Comment:
There's a possible race condition here if StartAsync and StopAsync are called close to each other.